maven-caching/src/main/java/org/apache/maven/caching/CacheControllerImpl.java

```diff
@@ -397,67 +415,126 @@ private RestoredArtifact restoredArtifact( org.apache.maven.artifact.Artifact pa
         return artifact;
     }
 
-    private Future<File> createDownloadTask( CacheResult cacheResult, CacheContext context, MavenProject project,
-                                             Artifact artifact )
-            throws ExecutionException, InterruptedException
+    //we might store in cache artifact which was build with previous version
+    //1.0-SNAPSHOT is kept in cache but real version of project is 2.0
+    //for pom packaging this is done automatically by maven but for jar and other there might be
+    //sensitive metadata with previous version. Versions mismatch could lead errors
+    private Path adjustArchiveArtifactVersion( MavenProject project, String originalArtifactVersion, Path artifactFile )
```

Comment:
i would move it to an utility class just to make it more testable

Reply:
> i would move it to an utility class just to make it easier testable

moved to CacheUtils